### PR TITLE
Cleanup XmlReader

### PR DESCRIPTION
**Pull request by kronic**
Cleanup XmlReader{Async}

### CODE REVIEW

File: src/libraries/System.Private.Xml/src/System/Xml/Core/XmlReader.cs [hunk 1]
```diff
@@ -87,33 +84,14 @@ public abstract partial class XmlReader : IDisposable
...
-                {
-                    return LocalName;
-                }
-                else
-                {
-                    return NameTable.Add(string.Concat(Prefix, ":", LocalName));
-                }
-            }
-        }
+        public virtual string Name => Prefix.Length == 0 ? LocalName : NameTable.Add(string.Concat(Prefix, ":", LocalName));
```
Comment (krwq) [inline src/libraries/System.Private.Xml/src/System/Xml/Core/XmlReader.cs:94]:
```suggestion
        public virtual string Name => Prefix.Length == 0 ? LocalName : NameTable.Add($"{Prefix}:{LocalName}");
```

File: src/libraries/System.Private.Xml/src/System/Xml/Core/XmlReader.cs [hunk 2]
```diff
@@ -1645,43 +1558,42 @@ internal string InternalReadContentAsString()
...
-                    case XmlNodeType.EndElement:
                     default:
                         goto ReturnContent;
                 }
-            } while ((this.AttributeCount != 0) ? this.ReadAttributeValue() : this.Read());
+            } while (AttributeCount != 0 ? ReadAttributeValue() : Read());
 
-        ReturnContent:
-            return (sb == null) ? value : sb.ToString();
+            ReturnContent:
```
Comment (krwq) [inline src/libraries/System.Private.Xml/src/System/Xml/Core/XmlReader.cs:1572]: I think previous indentation was correct (not 100% sure)

File: src/libraries/System.Private.Xml/src/System/Xml/Core/XmlReaderAsync.cs [hunk 3]
```diff
@@ -224,49 +218,49 @@ public virtual async Task<string> ReadInnerXmlAsync()
...
-                        xtw.WriteWhitespace(await this.GetValueAsync().ConfigureAwait(false));
+                        xtw.WriteWhitespace(await GetValueAsync().ConfigureAwait(false));
                         break;
                     case XmlNodeType.CDATA:
-                        xtw.WriteCData(this.Value);
+                        xtw.WriteCData(Value);
                         break;
                     case XmlNodeType.EntityReference:
-                        xtw.WriteEntityRef(this.Name);
+                        xtw.WriteEntityRef(Value);
```
Comment (krwq) [inline src/libraries/System.Private.Xml/src/System/Xml/Core/XmlReaderAsync.cs:246]: this is not cleanup

File: src/libraries/System.Private.Xml/src/System/Xml/Core/XmlReader.cs [hunk 4]
```diff
@@ -389,20 +325,18 @@ public virtual object ReadContentAs(Type returnType, IXmlNamespaceResolver? name
...
-                }
-                catch (FormatException e)
-                {
-                    throw new XmlException(SR.Xml_ReadContentAsFormatException, returnType.ToString(), e, this as IXmlLineInfo);
-                }
-                catch (InvalidCastException e)
-                {
-                    throw new XmlException(SR.Xml_ReadContentAsFormatException, returnType.ToString(), e, this as IXmlLineInfo);
-                }
+                return XmlUntypedConverter.Untyped.ChangeType(strContentValue, returnType, namespaceResolver ?? this as IXmlNamespaceResolver);
```
Comment (krwq) [inline src/libraries/System.Private.Xml/src/System/Xml/Core/XmlReader.cs:331]: please move any implementation changes to a separate PR

File: src/libraries/System.Private.Xml/src/System/Xml/Core/XmlReader.cs [hunk 5]
```diff
@@ -600,7 +534,7 @@ public virtual object ReadElementContentAs(Type returnType, IXmlNamespaceResolve
                 return value;
             }
 
-            return (returnType == typeof(string)) ? string.Empty : XmlUntypedStringConverter.Instance.FromString(string.Empty, returnType, namespaceResolver);
+            return returnType == typeof(string) ? string.Empty :  XmlUntypedConverter.Untyped.ChangeType(string.Empty, returnType, namespaceResolver);
```
Comment (krwq) [inline src/libraries/System.Private.Xml/src/System/Xml/Core/XmlReader.cs:537]: same

File: src/libraries/System.Private.Xml/src/System/Xml/Core/XmlReader.cs [hunk 6]
```diff
@@ -1,20 +1,17 @@
 // Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
-using System.Collections;
+using System.ComponentModel;
```
Comment (krwq) [inline src/libraries/System.Private.Xml/src/System/Xml/Core/XmlReader.cs:4]: why is new namespace needed?

**Review by krwq — CHANGES_REQUESTED** [review]
Please move any implementation changes to a separate PR

File: src/libraries/System.Private.Xml/src/System/Xml/Core/XmlReaderAsync.cs [hunk 7]
```diff
@@ -371,46 +365,46 @@ internal async Task<string> InternalReadContentAsStringAsync()
...
-                    case XmlNodeType.EndElement:
                     default:
                         goto ReturnContent;
                 }
-            } while ((this.AttributeCount != 0) ? this.ReadAttributeValue() : await this.ReadAsync().ConfigureAwait(false));
+            } while (AttributeCount != 0 ? ReadAttributeValue() : await ReadAsync().ConfigureAwait(false));
 
-        ReturnContent:
-            return (sb == null) ? value : sb.ToString();
+            ReturnContent:
```
Comment (jkotas) [inline src/libraries/System.Private.Xml/src/System/Xml/Core/XmlReaderAsync.cs:379]: This change goes against the coding style: https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/coding-style.md

When using labels (for goto), indent the label one less than the current indentation.

**Review by stephentoub — COMMENTED** [review]
Style changes generally look fine.  I agree with @krwq the functional changes need to be separated out.

**Comment by danmoseley** [comment]
That's a lot of semicolons, does that get us closer to enabling SA1106 globally?

**Comment by stephentoub** [comment]
> That's a lot of semicolons, does that get us closer to enabling SA1106 globally?

Semicolons?

Regardless, we won't be enabling SA1106:
https://github.com/dotnet/runtime/pull/50236#issuecomment-807851852

**Comment by danmoseley** [comment]
> Semicolons?

I was referring to the many empty statements that @kronic removed.

File: src/libraries/System.Private.Xml/src/System/Xml/Core/XmlReader.cs (hunk 6 above)
Comment (kronic) [reply src/libraries/System.Private.Xml/src/System/Xml/Core/XmlReader.cs:4]: [System.ComponentModel.EditorBrowsable(System.ComponentModel.EditorBrowsableState.Never)] ->         [EditorBrowsable(EditorBrowsableState.Never)]

**Comment by kronic** [comment]
@krwq done

File: src/libraries/System.Private.Xml/src/System/Xml/Core/XmlReader.cs [hunk 8]
```diff
@@ -600,7 +534,7 @@ public virtual object ReadElementContentAs(Type returnType, IXmlNamespaceResolve
                 return value;
             }
 
-            return (returnType == typeof(string)) ? string.Empty : XmlUntypedStringConverter.Instance.FromString(string.Empty, returnType, namespaceResolver);
+            return returnType == typeof(string) ? string.Empty : XmlUntypedStringConverter.Instance.FromString(string.Empty, returnType, namespaceResolver!);
```
Comment (krwq) [inline src/libraries/System.Private.Xml/src/System/Xml/Core/XmlReader.cs:537]: why is `!` needed now?

Comment (kronic) [reply src/libraries/System.Private.Xml/src/System/Xml/Core/XmlReader.cs:537]: Resharper gives warning Possible 'null' assignment to entity.
Next pr I plan to replace to XmlUntypedConverter.Untyped.ChangeType(string.Empty, returnType, namespaceResolver) as done in [(https://github.com/dotnet/runtime/blob/aa2af309a12d31dd68de9dc3646d77b6f25d9d40/src/libraries/System.Private.Xml/src/System/Xml/Core/XmlReaderAsync.cs#L106)]

Comment (kronic) [reply src/libraries/System.Private.Xml/src/System/Xml/Core/XmlReader.cs:537]: If necessary, can I remove it?

Comment (krwq) [reply src/libraries/System.Private.Xml/src/System/Xml/Core/XmlReader.cs:537]: @kronic, I assume if it wasn't needed before it shouldn't be now, especially namespaceResolver is non-nullable in that context. For next PR, let's talk about it in next PR, this is likely ok but prefer to think about it separately (also easier to track down if we regress something)

**Review by krwq — APPROVED** [review]
LGTM minus `!` comment

**Comment by kronic** [comment]
@krwq done

**Review by krwq — APPROVED** [review]

**Comment by krwq** [comment]
Thanks @kronic!

**Comment by kronic** [comment]
@krwq what are the problems of auto-merge

**Comment by krwq** [comment]
Seems CI leg is failing, it will not auto-merge until that pass. I'll see if the issue is related
